[Yaml] Added support for parsing PHP constants

HeahDude opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | TODO |

HeahDude committed:
[Yaml] added support for parsing PHP constants

HeahDude commented:
I was reading http://symfony.com/doc/current/components/dependency_injection/parameters.html#constants-as-parameters and felt sad about the note for yaml support.

Is this PR acceptable? Does this need a flag? Or to be reversible (which does not seem doable)?

There is no sense to handle such case with the dumper imho (unless it is passed as a string?? but what would be the use case?).

hhamon commented:
I'm -1 for such horrible hacks. YAML is not a suitable format to handle metada like describing a thing is a constant.

xabbuh commented:
Usually I would agree with @hhamon and I had the same feeling when reading the headline of this PR. But on the other hand @fabpot always wanted to have the Yaml component to support the things we need for Symfony (i.e. being able to configure the framework) instead of supporting the full YAML specs. Now, this is something that is supported by other configuration formats (i.e. PHP and XML), but which isn't supported for the YAML config format. Thus, I am :+1: for closing the gap.

xabbuh commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -562,6 +562,8 @@ private static function evaluateScalar($scalar, $flags, $references = array())
 
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
+                        return @constant(substr($scalar, 11));
```
I think we should to throw a `ParseException` here if the constant does not exist (or at least do that when the `PARSE_EXCEPTION_ON_INVALID_TYPE` flag is set).

HeahDude replied:
I agree, I first did it because this flag looks indeed appropriate. I don't know why I revert it.

So I'll push it quickly :)

xabbuh replied:
Well, now that I think of this again, I do not think we should evaluate the flag for this. Such an error imo should never be silenced.

HeahDude replied:
It's not an error, it's a warning, preventing to get `null`, remove the `@` and the last test fail because it gets `0.0` instead of `null`. That's why I silenced it (and removed the flag in the process without giving it a second thought).

xabbuh replied:
What I mean is that this usually indicates a mistake made by the developer when writing the YAML file.

xabbuh replied:
Hm, on the other hand they may have to deal with YAML files that did not write themself. But that's rather and argument to completely turn off parsing constants by default. Thus having the developer deliberately turn it on when wanting to parse constants.

HeahDude replied:
I guess getting `null` should be expected (a constant could be defined in an environment and not in another), the flag let the user decide it in a known and explicit way.

HeahDude replied:
Ok I now see you answered before me, I maintain my proposal though, but I'm not against yours.
 
Still I would rather not add a new flag for this, the goal here is to ease configuration ; currently you need to import an `xml` or a `php` file while `yaml` is very common (I think and always used it only).

So needing to add a flag does not seem to "keep it easy" while using `PARSE_EXCEPTION_ON_INVALID_TYPE` looks more like "keep doing as usual, just pass a constant if you want".

xabbuh replied:
@HeahDude Well, for the configuration of services the users would not have to do that. It's the responsibility of us to set up the parser properly in the `YamlFileLoader`. So we are more talking about someone who uses the parser in their own code.

HeahDude replied:
Understood! So I need to:

1. add a flag like `PARSE_PHP_CONSTANTS` in `Yaml` class

2. either throw a `ParseException` or return `null` when the constant does not exists
   maybe using `PARSE_EXCEPTION_ON_INVALID_TYPE`

3. add the flag `PARSE_PHP_CONSTANTS` to `YamlFileLoader` config.

Is that correct? What about point 2?

xabbuh replied:
I agree. And I think as you would then have to deliberately turn on the feature, throwing an exception when a constant does not exist is the right way.

HeahDude replied:
I'm almost there :)

Since the constant are not autoloaded (IIUC), one may expect to get `null` if a constant is not defined in an environment, so I guess we can take advantage of the triggered `E_WARNING` (ref http://php.net/manual/en/function.constant.php) as it will be shown as a silenced error in the profiler thanks to https://github.com/symfony/symfony/pull/16760.

Hence, I suggest we only trigger an exception when `PARSE_EXCEPTION_ON_INVALID_TYPE` flag is passed.

I already did such implementation but I've got an issue when the constant is not "inlined":
```yaml
foo:
    - '!php/const:PHP_INT_MAX' # does not work
bar: ['!php/const:PHP_INT_MAX'] # works
baz: { foo: '!php/const:PHP_INT_MAX' } # works
```

I think it comes from the `Parser` as I only changed the `Inline` class for now, any lead ?

Thanks.

iltar commented:
I would be really happy to see this implemented. I prefer yaml over xml by a lot for non-public packages, but not having constants is a big limitation to yaml :(

HeahDude committed:
[DI] added support for PHP constants in yaml configuration files

HeahDude commented:
Changelog:

 - f23dd9b addressed comments in original commit

 - 56297da added support in the DI `YamlFileLoader`

 - a58d960 added failing test cases

I was wrong in my previous comment it's also failing because the parser ignores special values when they are "nested", i.e: `foo: '!php/const:PHP_INT_MAX'`

xabbuh commented on src/Symfony/Component/DependencyInjection/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+3.2.0
+-----
+
+ * added support for PHP constants in yaml configuration files
```
YAML

xabbuh commented on src/Symfony/Component/DependencyInjection/composer.json:
```diff
@@ -19,7 +19,7 @@
     },
     "require-dev": {
-        "symfony/yaml": "~2.8|~3.0",
+        "symfony/yaml": "~2.8|~3.2",
```
Tests will fail on 2.8 now.

HeahDude replied:
Noted

xabbuh commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -562,6 +564,18 @@ private static function evaluateScalar($scalar, $flags, $references = array())
+                            }
+
+                            return $constant;
+                        }
```
This must throw an exception when constant support is disabled and `$exceptionOnInvalidType` is set (like we do for object support).

HeahDude replied:
Alright!

xabbuh commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -562,6 +564,18 @@ private static function evaluateScalar($scalar, $flags, $references = array())
+                        if (self::$constantSupport) {
+                            $constant = @constant(substr($scalar, 11));
+
+                            if (null === $constant && self::$exceptionOnInvalidType) {
```
This will fail the case where the constant was explicitly set to `null`.

HeahDude replied:
Arf! So the `E_WARNING` should be enough then, right?

HeahDude commented on src/Symfony/Component/Yaml/Tests/ParserTest.php:
```diff
@@ -1201,6 +1201,33 @@ public function getInvalidBinaryData()
+
+    public function testParseNestedPhpConstant()
+    {
+        $yaml = "foo: '!php/const:PHP_INT_MAX'";
```
What about this failing test?

xabbuh replied:
The tag will never be evaluated as you quoted the value (this is then just a simple string in YAML).

HeahDude replied:
Thanks, I've already understood my mistake and updated the tests, will push it as soon as I can :)

HeahDude commented:
Ok, comments addressed, tests are green :)

I guess the failure on low depth will have to wait for master to point on 3.2.

dunglas commented:
Supporting PHP constants would be nice but the proposed syntax `!php/const:PHP_INT_MAX` is not intuitive and looks ugly. Can we have another nicer syntax?

HeahDude commented:
@dunglas suggestions are welcome! It actually matches the object syntax `!php/object`, so I thought it would be both consistent and easy to learn. Have you something else in mind?

HeahDude commented:
Could be like float or binary `!! constant PHP_INT_MAX`.

stof commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -562,6 +564,15 @@ private static function evaluateScalar($scalar, $flags, $references = array())
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
+                        if (self::$constantSupport) {
+                            return @constant(substr($scalar, 11));
```
receiving null when a constant does not exist makes debugging very hard when you make a typo

HeahDude replied:
@stof there is a hidden conversation on a diff above between me and @xabbuh where we argue about this. It returns `null` but there is a warning logged as silenced error thanks to #16760.
Isn't that enough? Maybe we could also silence it in master for the `XmlLoader`?

stof replied:
well, IMO, we should not silence things here

HeahDude replied:
It would be consistent with the current `XmlLoader` indeed. But `null` will be return anyway with an `E_WARNING`.

nicolas-grekas replied:
We shouldn't we trigger an exception when the const is not defined? Would look better to me ATM

HeahDude replied:
Is there a way to distinguish `null` when assigned to a constant from when it's not defined?

nicolas-grekas replied:
yes, using `defined()`

HeahDude replied:
OK great! Then we should do it indeed. What do others think? @xabbuh @stof

stof commented:
> Could be like float or binary !! constant PHP_INT_MAX.

no it cannot. ``!!float`` is part of the Yaml standard. custom tags must star with a single ``!``, not 2

HeahDude commented:
Ok, so why not `!constant PHP_INT_MAX`? Although I'm not against the current syntax.

xabbuh commented:
I would not change the tag name. The `php:` part acts like a Namespace here which reduces the risk to conflict with custom tags from other libraries.

dunglas commented:
@xabbuh it makes perfectly sense but `!constant PHP_INT_MAX` is really nicer.

fabpot commented:
I'm against adding something that does not follow the YAML specification.

dunglas commented:
If I understand correctly the spec, both examples are valid: http://www.yaml.org/spec/1.2/spec.html#id2782728

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -22,6 +22,7 @@
 use Symfony\Component\Yaml\Exception\ParseException;
 use Symfony\Component\Yaml\Parser as YamlParser;
 use Symfony\Component\ExpressionLanguage\Expression;
+use Symfony\Component\Yaml\Yaml;
```
can you please move it one line up in the yaml group?

HeahDude replied:
Are you sure about this? The "good way" would be to move up the `Expression` line but as a CS fix it shouldn't be done in master, right?

nicolas-grekas replied:
let's group yaml lines now that we can (but we don't move existing lines to reduce merge conflicts)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -366,7 +367,7 @@ protected function loadFile($file)
 
         try {
-            $configuration = $this->yamlParser->parse(file_get_contents($file));
+            $configuration = $this->yamlParser->parse(file_get_contents($file), Yaml::PARSE_CONSTANT);
```
this means that the DI component not can't work with Yaml < 3.2, isn't it? So the composer.json file should get a conflict rule. Or: would it make sense to enable const parsing by default (and thus not require any change here?)

HeahDude replied:
Fixed

HeahDude replied:
@nicolas-grekas see https://github.com/symfony/symfony/pull/18626/files#diff-85ef95bd09276e850971f82e010b284dR32

nicolas-grekas commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -28,6 +28,7 @@ class Inline
     private static $exceptionOnInvalidType = false;
     private static $objectSupport = false;
     private static $objectForMap = false;
+    private static $constantSupport = false;
```
Are we sure we need a flag here? Couldn't we unconditionally enable this?

HeahDude replied:
@xabbuh what do you think?

xabbuh replied:
Depends on whether we want to throw exceptions when a constant is not defined. Imo the flag makes sense if we then throw an exception when a constant does not exist (which would make debugging a lot easier than when we just returned `null`). Though if we enabled this feature unconditionally, we would not be able to throw exceptions when the constant was not defined.

stof replied:
Throwing the exception already depends on the ``exceptionOnInvalidType`` flag currently

HeahDude replied:
Ok then, they are many options.

1. Parse constants by defaults:
 ```php
case 0 === strpos($scalar, '!php/const:'):
    if (defined($const = substr($scalar, 11))) {
        return constant($const);
    }
    if (self::$exceptionOnInvalidType) {
        throw new ParseException(sprintf('The PHP constant "%s" is not defined.', $const));
    }

    return;
 ```

2. Or using a flag as is:
 ```php
// use Symfony\Component\Yaml\Exception\RuntimeException;

case 0 === strpos($scalar, '!php/const:'):
    if (self::$constantSupport) {
        $const = substr($scalar, 11);
        if (defined($const)) {
            return constant($const);
        }

        throw new RuntimeException(sprintf('The constant "%s" is not defined.', $const));
    }
    if (self::$exceptionOnInvalidType) {
        throw new ParseException(sprintf('The string "%s" could not be parsed as constant. Have you forgotten to use "Yaml::PARSE_CONSTANT" flag?', $scalar));
    }

    return;
 ```

3. Suggestions?

For now I vote for option 2, since php objects are not parsed by default.

javiereguiluz replied:
My vote is for a flag too because this component should be as close to the Yaml spec as possible. `!php/const:` is a Symfony thing, not a Yaml thing.

xabbuh replied:
I agree with that. It will consistent with how we already treat object PHP support.

nicolas-grekas replied:
Fine by me, then [my comment above](https://github.com/symfony/symfony/pull/18626/files#r66247758) applies.

HeahDude commented on src/Symfony/Component/DependencyInjection/composer.json:
```diff
@@ -19,7 +19,7 @@
     },
     "require-dev": {
-        "symfony/yaml": "~2.8|~3.0",
+        "symfony/yaml": "~3.2",
```
@nicolas-grekas

>this means that the DI component not can't work with Yaml < 3.2, isn't it?

Isn't this line enough?

nicolas-grekas replied:
this line does apply only when testing the component, it has no effect when used as a dep of something else

fabpot commented:
@HeahDude Can you finish this PR?

HeahDude commented:
Yes! With option 2 from my comment above?

HeahDude commented:
Comments addressed. Waiting for tests to be green, needs review, thanks!

fabpot commented:
👍

HeahDude commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -366,7 +367,7 @@ protected function loadFile($file)
         try {
-            $configuration = $this->yamlParser->parse(file_get_contents($file));
+            $configuration = $this->yamlParser->parse(file_get_contents($file), Yaml::PARSE_CONSTANT);
         } catch (ParseException $e) {
```
The `Symfony\Component\Yaml\Exception\RuntimeException` might not be caught here, @xabbuh any ideas on how we should handle it?

nicolas-grekas replied:
@Haehnchen see https://github.com/symfony/symfony/pull/18626/files#r67163458

nicolas-grekas commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -581,6 +584,19 @@ private static function evaluateScalar($scalar, $flags, $references = array())
                         }
 
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
```
What about:
```php
                    case 0 === strpos($scalar, '!php/const:'):
                        if (self::$constantSupport && defined($const = substr($scalar, 11))) {
                            return constant($const);
                        }
                        if (self::$exceptionOnInvalidType) {
                            throw new ParseException(sprintf('The "%s" YAML string could not be parsed because the %s.', $scalar, self::$constantSupport ? ' "Yaml::PARSE_CONSTANT" flag was not set' : 'corresponding PHP constant was not defined'));
                        }

                        return; 
```

HeahDude replied:
@nicolas-grekas 👍

HeahDude replied:
@nicolas-grekas Working on it I think we should throw an exception when the constant is not defined even if `self::$exceptionOnInvalidType` is false, what do you think **@symfony/deciders**?

HeahDude replied:
In that case maybe I should only change the YAML `RuntimeException` to a `ParseException` so it's caught in the DI loader.

fabpot replied:
I think throwing an exception when a constant is not defined is a good idea.

nicolas-grekas commented:
Status: needs work

HeahDude commented:
Updated. Status: Needs Review

HeahDude commented:
Status: Needs Review

fabpot commented:
👍 (failures not related)

xabbuh commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -581,6 +583,19 @@ private static function evaluateScalar($scalar, $flags, $references = array())
+                            throw new ParseException(sprintf('The constant "%s" is not defined.', $const));
+                        }
+                        if (self::$exceptionOnInvalidType) {
+                            throw new ParseException(sprintf('The string "%s" could not be parsed as constant. Have you forgotten to pass "Yaml::PARSE_CONSTANT" flag to the parser?', $scalar));
```
[...] pass **the** [...]

xabbuh replied:
[...] be parsed as **a** constant.

xabbuh commented on src/Symfony/Component/Yaml/Tests/InlineTest.php:
```diff
@@ -36,6 +36,34 @@ public function testParseWithMapObjects($yaml, $value)
+    {
+        $expectedMessage = 'The constant "WRONG_CONSTANT" is not defined';
+
+        $this->setExpectedException('\Symfony\Component\Yaml\Exception\ParseException', $expectedMessage);
```
Can you use the `@expectedException` annotation instead?

xabbuh commented on src/Symfony/Component/Yaml/Tests/InlineTest.php:
```diff
@@ -36,6 +36,34 @@ public function testParseWithMapObjects($yaml, $value)
+    {
+        $expectedMessage = '#The string "!php/const:PHP_INT_MAX" could not be parsed as constant.*#';
+
+        $this->setExpectedExceptionRegExp('\Symfony\Component\Yaml\Exception\ParseException', $expectedMessage);
```
same here

xabbuh commented on src/Symfony/Component/Yaml/Tests/InlineTest.php:
```diff
@@ -442,6 +470,16 @@ public function getTestsForParseWithMapObjects()
+            array('[!php/const:PHP_INT_MAX]', array(PHP_INT_MAX)),
+            array('{ foo: !php/const:PHP_INT_MAX }', array('foo' => PHP_INT_MAX)),
+        );
+    }
```
I would move the data provider just below the `testParsePhpConstants` test.

HeahDude replied:
Done

xabbuh commented on src/Symfony/Component/Yaml/Tests/ParserTest.php:
```diff
@@ -1351,6 +1351,30 @@ public function parserThrowsExceptionWithCorrectLineNumberProvider()
+        $this->setExpectedExceptionRegExp('\Symfony\Component\Yaml\Exception\ParseException', $expectedMessage);
+
+        $this->parser->parse('foo: !php/const:PHP_INT_MAX', Yaml::PARSE_EXCEPTION_ON_INVALID_TYPE);
+    }
```
Not sure if it's really necessary to duplicate all the tests from the `Inline` class here.

HeahDude replied:
@xabbuh not "all" so we'd just keep the one without exception?

HeahDude replied:
@xabbuh or do you mean I can remove all of them?

xabbuh replied:
I would simply revert the changes in this file.

fabpot commented:
@HeahDude Can you take @xabbuh comments into account?

fabpot commented:
Thank you @HeahDude.

teohhanhui commented:
It's unfortunate that we did not consider it important to adhere to the YAML spec: http://www.yaml.org/spec/1.2/spec.html#tag/local/
